File: spring-data-requery/src/main/java/org/springframework/data/requery/repository/support/SimpleRequeryRepository.java
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.repository.support;

import io.requery.meta.Attribute;
import io.requery.query.Condition;
import io.requery.query.LogicalCondition;
import io.requery.query.NamedExpression;
import io.requery.query.OrderingExpression;
import io.requery.query.Result;
import io.requery.query.Return;
import io.requery.query.Tuple;
import io.requery.query.element.QueryElement;
import io.requery.query.function.Count;
import io.requery.sql.EntityDataStore;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.dao.IncorrectResultSizeDataAccessException;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.requery.core.RequeryOperations;
import org.springframework.data.requery.repository.query.QueryByExampleBuilder;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;

import static org.springframework.data.requery.utils.RequeryUtils.applyPageable;
import static org.springframework.data.requery.utils.RequeryUtils.applySort;
import static org.springframework.data.requery.utils.RequeryUtils.foldConditions;
import static org.springframework.data.requery.utils.RequeryUtils.getKeyExpression;
import static org.springframework.data.requery.utils.RequeryUtils.getOrderingExpressions;
import static org.springframework.data.requery.utils.RequeryUtils.unwrap;

/**
 * Default implementation of the {@link org.springframework.data.repository.CrudRepository} interface.
 * This will offer you a more sophisticated interface than the plain {@link EntityDataStore}.
 *
 * @author debop
 * @since 18. 6. 4
 */
@SuppressWarnings("ConstantConditions")
@Slf4j
@Repository
@Transactional(readOnly = true)
public class SimpleRequeryRepository<T, ID> implements RequeryRepositoryImplementation<T, ID> {

    @Getter
    @Nonnull private final RequeryOperations operations;
    @Nonnull private final RequeryEntityInformation<T, ID> entityInformation;

    @Nullable private final Class<T> domainClass;
    @Nonnull private final String domainClassName;

    @Nullable private CrudMethodMetadata crudMethodMetadata;

    public SimpleRequeryRepository(@Nonnull final RequeryEntityInformation<T, ID> entityInformation,
                                   @Nonnull final RequeryOperations operations) {
        log.debug("Create SimpleRequeryRepository. domainClass={}", entityInformation.getJavaType());

        this.entityInformation = entityInformation;
        this.domainClass = entityInformation.getJavaType();
        this.domainClassName = (domainClass != null) ? domainClass.getSimpleName() : "Unknown";
        this.operations = operations;
    }

    @Override
    public void setRepositoryMethodMetadata(@Nullable final CrudMethodMetadata crudMethodMetadata) {
        this.crudMethodMetadata = crudMethodMetadata;
    }

    @Transactional
    @Override
    @Nonnull
    public <S extends T> S insert(@Nonnull final S entity) {
        return operations.insert(entity);
    }

    @Transactional
    @Override
    @Nonnull
    public <S extends T, K> K insert(@Nonnull final S entity, @Nonnull final Class<K> keyClass) {
        return operations.insert(entity, keyClass);
    }

    @Transactional
    @Override
    @Nonnull
    public <S extends T> List<S> insert(@Nonnull final Iterable<S> entities) {
        return operations.insertAll(entities);
    }

    @Transactional
    @Override
    @Nonnull
    public <S extends T, K> List<K> insert(@Nonnull final Iterable<S> entities, @Nonnull final Class<K> keyClass) {
        return operations.insertAll(entities, keyClass);
    }

    @Transactional
    @Override
    @Nonnull
    public <S extends T> S upsert(@Nonnull final S entity) {
        return operations.upsert(entity);
    }

    @Transactional
    @Override
    @Nonnull
    public <S extends T> List<S> upsertAll(@Nonnull final Iterable<S> entities) {
        return operations.upsertAll(entities);
    }

    @Nonnull
    public <S extends T> S refresh(@Nonnull final S entity) {
        return operations.refresh(entity);
    }

    @SuppressWarnings({ "unchecked", "NullableProblems" })
    @Nonnull
    public <S extends T> List<S> refresh(@Nonnull final Iterable<S> entities, final Attribute<S, ?>... attributes) {
        return operations.refresh(entities, attributes);
    }

    @Nonnull
    public <S extends T> S refreshAll(@Nonnull final S entity) {
        return operations.refreshAllProperties(entity);
    }

    @Transactional
    @Override
    public void deleteInBatch(@Nonnull final Iterable<T> entities) {
        operations.deleteAll(entities);
    }

    @Transactional
    @Override
    public int deleteAllInBatch() {
        return operations.delete(domainClass).get().value();
    }

    @Override
    public T getOne(@Nonnull final ID id) {
        return operations.findById(domainClass, id);
    }

    @Nonnull
    @Override
    public List<T> findAll(@Nonnull final Sort sort) {

        log.debug("Find all {} with sort, sort={}", domainClassName, sort);

        if (sort.isSorted()) {
            OrderingExpression<?>[] orderingExprs = getOrderingExpressions(domainClass, sort);

            if (orderingExprs.length > 0) {
                return operations
                    .select(domainClass)
                    .orderBy(orderingExprs)
                    .get()
                    .toList();
            }
        }

        return operations
            .select(domainClass)
            .get()
            .toList();
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public Page<T> findAll(@Nonnull final Pageable pageable) {

        log.debug("Find all {} with paging, pageable={}", domainClassName, pageable);

        if (pageable.isPaged()) {
            QueryElement<? extends Result<T>> query = (QueryElement<? extends Result<T>>)
                applyPageable(domainClass,
                              (QueryElement<? extends Result<T>>) operations.select(domainClass),
                              pageable);
            List<T> content = query.get().toList();
            long total = operations.count(domainClass).get().value().longValue();

            return new PageImpl<>(content, pageable, total);
        } else {
            List<T> content = operations
                .select(domainClass)
                .get()
                .toList();
            return new PageImpl<>(content);
        }
    }

    @Transactional
    @Nonnull
    @Override
    public <S extends T> S save(@Nonnull final S entity) {
        return operations.upsert(entity);
    }

    @Transactional
    @Nonnull
    @Override
    public <S extends T> List<S> saveAll(@Nonnull final Iterable<S> entities) {
        return operations.upsertAll(entities);
    }

    @Nonnull
    @Override
    public Optional<T> findById(@Nonnull final ID id) {
        return Optional.ofNullable(operations.findById(domainClass, id));
    }

    @SuppressWarnings("unchecked")
    @Override
    public boolean existsById(@Nonnull final ID id) {
        NamedExpression<ID> keyExpr = (NamedExpression<ID>) getKeyExpression(domainClass);

        Tuple result = operations
            .select(Count.count(domainClass))
            .where(keyExpr.eq(id))
            .get()
            .first();

        return result.<Integer>get(0) > 0;
    }

    @Nonnull
    @Override
    public List<T> findAll() {
        return operations.findAll(domainClass);
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public List<T> findAllById(@Nonnull final Iterable<ID> ids) {
        HashSet<ID> idSet = new HashSet<>();
        for (ID id : ids) {
            idSet.add(id);
        }
        NamedExpression<ID> keyExpr = (NamedExpression<ID>) getKeyExpression(domainClass);

        return operations
            .select(domainClass)
            .where(keyExpr.in(idSet))
            .get()
            .toList();
    }

    @Override
    public long count() {
        return operations
            .count(domainClass)
            .get()
            .value()
            .longValue();
    }

    @SuppressWarnings("unchecked")
    @Transactional
    @Override
    public void deleteById(@Nonnull final ID id) {
        log.debug("Delete {} entity by id. id={}", domainClassName, id);

        NamedExpression<ID> keyExpr = (NamedExpression<ID>) getKeyExpression(domainClass);

        Integer deletedCount = operations
            .delete(domainClass)
            .where(keyExpr.eq(id))
            .get()
            .value();

        log.debug("Deleted entity={}, count={}", domainClassName, deletedCount);
    }

    @Transactional
    @Override
    public void delete(@Nonnull final T entity) {
        operations.delete(entity);
    }

    @Transactional
    @Override
    public void deleteAll(@Nonnull final Iterable<? extends T> entities) {
        operations.deleteAll(entities);
    }

    @Transactional
    @Override
    public void deleteAll() {
        log.debug("Delete All entities. entity name={} ...", domainClassName);

        Integer deletedCount = operations.delete(domainClass).get().value();

        log.debug("Delete All entities. entity name={}, deleted count={}", domainClassName, deletedCount);
    }

    @Nonnull
    @Override
    public <S extends T> Optional<S> findOne(@Nonnull final Example<S> example) {
        return Optional.ofNullable(buildQueryByExample(example).get().firstOrNull());
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public <S extends T> List<S> findAll(@Nonnull final Example<S> example) {
        return buildQueryByExample(example).get().toList();
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public <S extends T> List<S> findAll(@Nonnull final Example<S> example, @Nonnull final Sort sort) {
        QueryElement<?> query = applySort(domainClass,
                                          unwrap(buildQueryByExample(example)),
                                          sort);
        return ((QueryElement<? extends Result<S>>) query).get().toList();
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public <S extends T> Page<S> findAll(@Nonnull final Example<S> example, @Nonnull final Pageable pageable) {

        QueryElement<?> baseQuery = unwrap(buildQueryByExample(example));
        long total = count((QueryElement<? extends Result<T>>) baseQuery);

        QueryElement<?> query = applyPageable(domainClass,
                                              unwrap(buildQueryByExample(example)),
                                              pageable);
        return new PageImpl<>(((QueryElement<? extends Result<S>>) query).get().toList(),
                              pageable,
                              total);
    }

    @SuppressWarnings("unchecked")
    @Override
    public <S extends T> long count(@Nonnull Example<S> example) {
        return count((QueryElement<? extends Result<T>>) buildQueryByExample(example));
    }

    @SuppressWarnings("unchecked")
    @Override
    public <S extends T> boolean exists(@Nonnull Example<S> example) {
        return buildQueryByExample(example).limit(1).get().firstOrNull() != null;
    }

    @SuppressWarnings("unchecked")
    private <S extends T> QueryElement<? extends Result<S>> buildQueryByExample(@Nonnull Example<S> example) {
        QueryElement<? extends Result<S>> root = (QueryElement<? extends Result<S>>) unwrap(operations.select(domainClass));
        return QueryByExampleBuilder.applyExample(root, example);
    }

    @Nonnull
    @Override
    public Optional<T> findOne(@Nonnull final Return<? extends Result<T>> whereClause) {
        List<T> list = whereClause.get().toList();
        if (list.size() == 1) {
            return Optional.ofNullable(list.get(0));
        } else if (list.isEmpty()) {
            return Optional.empty();
        } else {
            throw new IncorrectResultSizeDataAccessException(1, list.size());
        }
    }

    @Nonnull
    @Override
    public List<T> findAll(@Nonnull final Return<? extends Result<T>> whereClause) {
        return whereClause.get().toList();
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public Page<T> findAll(@Nonnull final QueryElement<? extends Result<T>> whereClause,
                           @Nonnull final Pageable pageable) {

        int total = count(whereClause);

        Return<?> query = applyPageable(domainClass, whereClause, pageable);
        List<T> contents = ((Return<? extends Result<T>>) query).get().toList();

        return new PageImpl<>(contents, pageable, total);
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public List<T> findAll(@Nonnull final Iterable<Condition<T, ?>> conditions,
                           @Nonnull final Sort sort) {

        LogicalCondition<T, ?> condition = foldConditions(conditions);
        QueryElement<?> query = unwrap(operations.select(domainClass).where(condition));
        return ((QueryElement<? extends Result<T>>) applySort(domainClass, query, sort)).get().toList();
    }

    @SuppressWarnings("unchecked")
    @Nonnull
    @Override
    public List<T> findAll(@Nonnull final Iterable<Condition<T, ?>> conditions) {

        LogicalCondition<T, ?> condition = foldConditions(conditions);
        return operations.select(domainClass).where(condition).get().toList();
    }

    @SuppressWarnings("unchecked")
    @Override
    public int count(@Nonnull final QueryElement<? extends Result<T>> whereClause) {
        return getOperations().count(domainClass, whereClause);
    }

    @Override
    public boolean exists(@Nonnull final QueryElement<? extends Result<T>> whereClause) {
        return getOperations().exists(domainClass, whereClause);
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/repository/support/SimpleRequeryRepository.kt
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.kotlin.repository.support

import io.requery.meta.Attribute
import io.requery.query.Condition
import io.requery.query.Result
import io.requery.query.Return
import io.requery.query.element.QueryElement
import io.requery.query.function.Count
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.IncorrectResultSizeDataAccessException
import org.springframework.data.domain.Example
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.requery.kotlin.applyPageable
import org.springframework.data.requery.kotlin.applySort
import org.springframework.data.requery.kotlin.buildQueryElement
import org.springframework.data.requery.kotlin.core.RequeryOperations
import org.springframework.data.requery.kotlin.foldConditions
import org.springframework.data.requery.kotlin.getAsResultEntity
import org.springframework.data.requery.kotlin.getKeyExpression
import org.springframework.data.requery.kotlin.unwrap
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import java.util.*
import kotlin.reflect.KClass

/**
 *
 * [org.springframework.data.requery.kotlin.repository.RequeryRepository] 의 기본 구현체
 *
 * @author debop
 */
@Suppress("UNCHECKED_CAST")
@Repository
@Transactional(readOnly = true)
class SimpleRequeryRepository<E : Any, ID : Any> @Autowired constructor(
    final val entityInformation: RequeryEntityInformation<E, ID>,
    override val operations: RequeryOperations
) : RequeryRepositoryImplementation<E, ID> {

    companion object {
        private val log = KotlinLogging.logger { }
    }

    final override val domainKlass: KClass<E> = entityInformation.kotlinType
    final override val domainClass: Class<E> = entityInformation.javaType
    final val domainClassName: String = domainClass.simpleName ?: "Unknown"

    private var crudMethodMetadata: CrudMethodMetadata? = null

    init {
        log.info { "Create SimpleRequeryRepository for domainClass=${domainKlass.simpleName}" }
    }

    override fun setRepositoryMethodMetadata(crudMethodMetadata: CrudMethodMetadata?) {
        this.crudMethodMetadata = crudMethodMetadata
    }

    fun select(): QueryElement<out Result<E>> = operations.select(domainKlass).unwrap()

    override fun findAll(): List<E> = operations.findAll(domainKlass)

    override fun findAll(sort: Sort): List<E> {
        return select()
            .applySort(domainClass, sort)
            .get()
            .toList()
    }

    override fun <S : E> findAll(example: Example<S>): List<S> {
        return example.buildQueryElement(operations, domainKlass as KClass<S>).get().toList()
    }

    override fun <S : E> findAll(example: Example<S>, sort: Sort): List<S> {
        return example
            .buildQueryElement(operations, domainKlass)
            .applySort(domainClass, sort)
            .get()
            .toList()
    }

    override fun findAll(pageable: Pageable): Page<E> {
        log.trace { "Fild all $domainClassName with paging. pageable=$pageable" }

        return when {
            pageable.isPaged -> {
                val content = select().applyPageable(domainClass, pageable).getAsResultEntity<E>().toList()
                val totals = operations.count(domainKlass).get().value().toLong()

                PageImpl(content, pageable, totals)
            }
            else -> {
                val content = select().get().toList()
                PageImpl(content)
            }
        }
    }

    override fun <S : E> findAll(example: Example<S>, pageable: Pageable): Page<S> {

        log.trace { "Find all [$domainClass] with paging. pageable=$pageable" }

        val queryElement = example.buildQueryElement(operations, domainKlass).unwrap()

        return when {
            pageable.isPaged -> {
                val totals = count(example)
                val contents = queryElement.applyPageable(domainClass, pageable).get().toList()

                PageImpl<S>(contents, pageable, totals)
            }
            else -> {
                val contents = queryElement.get().toList()
                PageImpl<S>(contents)
            }
        }
    }

    override fun findAll(filter: Return<out Result<E>>): List<E> {
        return filter.get().toList()
    }

    override fun findAll(filter: QueryElement<out Result<E>>, pageable: Pageable): Page<E> {
        return when {
            pageable.isPaged -> {
                val totals = count(filter)
                val contents = filter.applyPageable(domainClass, pageable).get().toList()
                PageImpl<E>(contents, pageable, totals)
            }
            else -> PageImpl<E>(findAll(filter))
        }
    }

    override fun findAll(conditions: Iterable<Condition<E, *>>): List<E> {
        val whereClause = conditions.foldConditions()

        return whereClause?.let {
            select().where(it).get().toList()
        } ?: emptyList()
    }

    override fun findAll(conditions: Iterable<Condition<E, *>>, pageable: Pageable): Page<E> {
        return when {
            pageable.isPaged -> {
                val whereClause = conditions.foldConditions()
                val baseQuery = whereClause?.let {
                    select().where(it).unwrap()
                } ?: select()

                val contents = baseQuery.unwrap().applyPageable(domainClass, pageable).get().toList()
                PageImpl<E>(contents, pageable, count(baseQuery))
            }
            else ->
                PageImpl<E>(findAll(conditions))
        }
    }

    override fun findAll(conditions: Iterable<Condition<E, *>>, sort: Sort): List<E> {
        return select()
            .where(conditions.foldConditions())
            .applySort(domainClass, sort)
            .get()
            .toList()
    }

    override fun findAllById(ids: Iterable<ID>): List<E> {

        log.trace { "Find all by id. ids=$ids" }

        val keyExpr = domainClass.getKeyExpression<ID>()

        return operations
            .select(domainKlass)
            .where(keyExpr.`in`(ids.toSet()))
            .get()
            .toList()
    }

    override fun <S : E> saveAll(entities: Iterable<S>): List<S> {
        return operations.upsertAll(entities)
    }

    override fun insert(entity: E): E = operations.insert(entity)

    override fun <K : Any> insert(entity: E, keyClass: KClass<K>): K {
        return operations.insert(entity, keyClass).also {
            log.trace { "Insert entity, new key=$it" }
        }
    }

    override fun insertAll(entities: Iterable<E>): List<E> {
        return operations.insertAll(entities)
    }

    override fun <K : Any> insertAll(entities: Iterable<E>, keyClass: KClass<K>): List<K> {
        return operations.insertAll(entities, keyClass).also {
            log.trace { "Insert entities, new keys=$it" }
        }
    }

    override fun upsert(entity: E): E {
        return operations.upsert(entity)
    }

    override fun upsertAll(entities: Iterable<E>): List<E> {
        return operations.upsertAll(entities)
    }

    override fun refresh(entity: E): E {
        return operations.refresh(entity)
    }

    override fun refreshAllProperties(entity: E): E {
        return operations.refreshAllProperties(entity)
    }

    override fun refreshAll(entities: Iterable<E>, vararg attributes: Attribute<E, *>): List<E> {
        return operations.refreshAll(entities, *attributes)
    }

    override fun refreshAllEntities(entities: Iterable<E>, vararg attributes: Attribute<E, *>): List<E> {
        return operations.refreshAllEntities(entities, *attributes)
    }

    override fun deleteInBatch(entities: Iterable<E>) {
        operations.deleteAll(entities)
    }

    override fun deleteAllInBatch(): Int {
        return operations.deleteAll(domainKlass)
    }

    override fun getOne(id: ID): E? {
        return operations.findById(domainKlass, id)
    }

    override fun <S : E> save(entity: S): S {
        return operations.upsert(entity)
    }

    override fun deleteById(id: ID) {
        log.trace { "Delete $domainClassName by id [$id]" }

        val keyExpr = domainClass.getKeyExpression<ID>()

        val deletedCount = operations
            .delete(domainKlass)
            .where(keyExpr.eq(id))
            .get()
            .value()

        log.trace { "Delete $domainClassName by id [$id]. deleted count=$deletedCount" }
    }

    override fun deleteAll(entities: MutableIterable<E>) {
        operations.deleteAll(entities)
    }

    override fun deleteAll() {
        log.debug { "Delete all entities ... domainClass=$domainClassName" }
        operations.deleteAll(domainKlass)
    }

    override fun count(): Long {
        return operations.count(domainKlass).get().value().toLong()
    }

    override fun <S : E> count(example: Example<S>): Long {

        return count(example.buildQueryElement(operations, domainKlass as KClass<S>) as QueryElement<out Result<E>>)
    }

    override fun count(queryElement: QueryElement<out Result<E>>): Long {
        return operations.count(domainKlass, queryElement).toLong()
    }

    override fun existsById(id: ID): Boolean {
        val keyExpr = domainClass.getKeyExpression<ID>()

        val tuple = operations
            .select(Count.count(domainKlass.java))
            .where(keyExpr.eq(id))
            .get()
            .firstOrNull()

        return tuple.get<Int>(0) > 0
    }

    override fun findById(id: ID): Optional<E> {
        return Optional.ofNullable(operations.findById(domainKlass, id))
    }

    override fun delete(entity: E) {
        log.trace { "Delete entity. entity=$entity" }
        operations.delete(entity)
    }

    override fun <S : E> findOne(example: Example<S>): Optional<S> {
        val entity = example.buildQueryElement(operations, domainKlass)
            .limit(1)
            .get()
            .firstOrNull()

        return Optional.ofNullable(entity)
    }

    override fun findOne(filter: Return<out Result<E>>): Optional<E> {
        val count = count(filter.unwrap()).toInt()
        if(count > 1) {
            throw IncorrectResultSizeDataAccessException(1, count)
        }
        return Optional.ofNullable(filter.get().firstOrNull())
    }

    override fun <S : E> exists(example: Example<S>): Boolean {
        return example
            .buildQueryElement(operations, domainKlass)
            .limit(1)
            .get()
            .firstOrNull() != null
    }

    override fun exists(queryElement: QueryElement<out Result<E>>): Boolean {
        return operations.exists(domainKlass, queryElement)
    }

}